Add Psalm - Level 4

DanielBadura opened the pull request:
I added psalm with level 4.
Still missing is the adjustment for travis. tbd.

Ocramius commented on psalm.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0"?>
+<psalm
+    errorLevel="4"
```
suggest dropping:

 * `errorLevel`
 * `resolveFromConfigFile`

And then adding:

 * `totallyType="true"`
 * a baseline file

Ocramius commented on src/ComposerRequireChecker/ASTLocator/LocateASTFromFiles.php:
```diff
@@ -25,9 +25,7 @@ public function __construct(Parser $parser, ?ErrorHandler $errorHandler)
-     * @param Traversable|string[] $files
-     *
-     * @return Traversable|array[] a series of AST roots, one for each given file
+     * @return Traversable a series of AST roots, one for each given file
```
We should document the `Traversable` input and output parameters - leaving them as just `Traversable` is not acceptable anymore

Ocramius commented on src/ComposerRequireChecker/DefinedSymbolsLocator/LocateDefinedSymbolsFromASTRoots.php:
```diff
@@ -10,7 +10,7 @@
 {
     /**
-     * @param Traversable|array[] $ASTs a series of AST roots
+     * @param Traversable $ASTs a series of AST roots
```
`Traversable<???, array<???>>`

Ocramius commented on src/ComposerRequireChecker/FileLocator/LocateFilesByGlobPattern.php:
```diff
@@ -11,10 +11,10 @@ class LocateFilesByGlobPattern
      * @param string[] $globPatterns a list of glob patterns to find files in
      * @param string $rootDir the root directory that should be used when patterns are relative paths
-     * @return iterable|string[] the files found by the given glob patterns
+     * @return Generator the files found by the given glob patterns
```
I suggest keeping these as `iterable`: don't want people to call `Generator` methods on this result (especially not `Generator#send()` :D )

Let's also document that this is an `iterable<???, string>`

Ocramius commented on src/ComposerRequireChecker/UsedSymbolsLocator/LocateUsedSymbolsFromASTRoots.php:
```diff
@@ -10,7 +10,7 @@
 {
     /**
-     * @param Traversable|array[] $ASTs a series of AST roots
+     * @param Traversable $ASTs a series of AST roots
```
Let's document this `Traversable<???, ???>` type

Ocramius commented on src/ComposerRequireChecker/ASTLocator/LocateASTFromFiles.php:
```diff
@@ -25,7 +26,9 @@ public function __construct(Parser $parser, ?ErrorHandler $errorHandler)
-     * @return Traversable a series of AST roots, one for each given file
+     * @param Traversable<string> $files
+     *
+     * @return Traversable<int, array<Stmt|null>|mixed> a series of AST roots, one for each given file
```
`mixed` seems borked here: can we avoid it?

Ocramius commented on src/ComposerRequireChecker/DefinedSymbolsLocator/LocateDefinedSymbolsFromASTRoots.php:
```diff
@@ -3,14 +3,15 @@
 {
     /**
-     * @param Traversable $ASTs a series of AST roots
+     * @param Traversable<int, array<Node|null>|mixed> $ASTs a series of AST roots
```
`mixed` also here defeats the other types

DanielBadura replied:
well see my comment 🗡️ 
You are just to fast with the review 😆

Ocramius commented on src/ComposerRequireChecker/UsedSymbolsLocator/LocateUsedSymbolsFromASTRoots.php:
```diff
@@ -3,14 +3,15 @@
 {
     /**
-     * @param Traversable $ASTs a series of AST roots
+     * @param Traversable<int, array<Node|null>|mixed> $ASTs a series of AST roots
```
`mixed` kinda defeats the purpose here

DanielBadura commented:
Thanks for the review! I did not want to change much of the types in this pr but i did it now. But i have some mixed right now that i want to get rid of. I will look into it.

Maybe the thing with the errorLevel & strictTypes in another pr and this is just the first step to introduce psalm? :)

Ocramius commented on src/ComposerRequireChecker/GeneratorUtil/ComposeGenerators.php:
```diff
@@ -2,11 +2,11 @@
 final class ComposeGenerators
 {
-    public function __invoke(Generator ...$generators): Generator
+    public function __invoke(Traversable ...$generators): Traversable
```
This could be `Traversable<TKey, TValue>` as input, and `Traversable<int, TValue>` as output

Ocramius commented:
@DanielBadura seems like an acceptable tradeoff

DanielBadura commented:
@Ocramius so now i got rid of `mixed` and even `null`. If the DocType on `PhpParser\Parser::parse` is correct, then `null` seems like a parse error, so i introduced a `RuntimeException` for this case. Is this to hard? Do we need to support the `null` case?

The other thing is that phpstan was correct to say that `is_string` in `UsedSymbolCollector::recordFunctionReturnTypeUsage` is always `false` since there should never be a `string` in it. I removed it and the 2 UnitTest that were  checking on it. What do you think?

Ocramius commented:
> Is this to hard? Do we need to support the `null` case?

Don't think so - the codebase should be parseable before being analysed :)

 > I removed it and the 2 UnitTest that were checking on it. What do you think?

Should be correct: impossible type intersection, and one less conditional :+1:

DanielBadura commented:
I added a new TestCase for the new introduced `RuntimeError`.
I mocked the parser to achieve the `null` result because i could not get the right input for this. Maybe you know what i need to do to get the parser to return `null` 😆 

Also added psalm now to travis.
Form my side this is now ready, if you want to delay https://github.com/maglnet/ComposerRequireChecker/pull/194#discussion_r461415282 for another pr.

Ocramius commented:
Thanks @DanielBadura!
